design/baremetal-operator/external-introspection.md

```diff
@@ -0,0 +1,116 @@
+# Support external introspection
+
+## Status
+
+implementable
+
+## Summary
+
+A declarative API is proposed to disable inspection of a BareMetalHost
+and optionally allow external sources of inspection data to update the
+hardware status data.
+
+## Motivation
+
+Related to the work to enable booting of a
+[LiveImage](https://github.com/metal3-io/metal3-docs/pull/150),
+there is the requirement to optionally disable inspection on initial
+registration of a BareMetalHost (so that the live image can boot
+more quickly, ref user stories below).
+
+### Goals
+
+- A declarative API to disable inspection on BMH registration
+- Provide an interface to update hardware status data when inspection is
+  performed by some means other than the regular BMH flow at an arbitrary time
+  after BMH creation.
+- An interface that's consistent with existing proposals, in particular the
+  [re-inspection interface](https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md)
+
+### Non-Goals
+
+- There are no plans for any mechanism to trigger or consume data from any out
+  of band inspection process, other than providing data via a status annotation.
+
+## Proposal
+
+### Disable inspection proposal
+
+To align with the re-inspection API proposal, the `inspect.metal3.io`
+annotation will be reused, but with an optional additional suffix, similar
+to that described for the [reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/reboot-interface.md)
+
+The optional `inspect.metal3.io/disable` annotation will be used to describe
+the situation where we wish to disable the default inspection behavior.
+
+When the BMO finds this annotation, it will skip performing inspection
+during the
+[Inspecting state](https://github.com/metal3-io/baremetal-operator/blob/master/docs/BaremetalHost_ProvisioningState.png)
+
+In this case (similar to the reboot interface), when the suffixed form of
+`inspect.metal3.io/disable` is found, it will never be removed by the BMO.
+
+If both inspect.metal3.io and inspect.metal3.io/disable forms are specified,
+the `inspect.metal3.io/disable` form will take priority and the
+`inspect.metal3.io` annotation will be ignored.
```

Comment:
Possible alternative: use the same `inspect.metal3.io` annotation and decide what to do based on the content (this way you can never have a conflict), e.g.:
* (empty): re-inspection as proposed
* `disable`: don't inspect
* `{...}` (HardwareStatus in json format): replace HardwareStatus with this data

Reply:
I like the way this avoids conflicts. Are there any precedents of using annotation content of different types in other APIS?

Reply:
I couldn't find any examples of this kind of interface, but it seems reasonable to me - the only slightly weird thing is combining one key to influence both the action (re-inspect, disable) and the data, what about separating them, e.g:

```
inspect.metal3.io/disable
inspect.metal3.io/data: <data in json format>
```

Then the bare `inspect.metal3.io` can be interpreted as re-triggering inspection, and we've got the ability to expand the interface in future if needed?

Reply:
Using two keys sounds cleaner. What about collapsing everything in just one key, and specify the action as part of the json payload (in additon to the data)? It could help with the unmarshalling and there will be just one key

Reply:
@andfasano I was thinking it would be cleaner to keep the action and data separate, it'll be simpler to just unmashall whatever is in `inspect.metal3.io/data` and put that into `status->hardware`?

What is the benefit of having only one key, vs using the `/subkey` approach?

Reply:
Just to slightly reduce the complexity (by mistake, one could define the `/data` subkey without the master action key, maybe it could be more difficult when both the things are defined as part of the json payload), but maybe it could be too much overhead in some cases. So fine for me to have a couple of subkeys, one dedicated for just the data when required.

Reply:
Ok thanks - personally I'd prefer to avoid merging the action with the json data, so IMO we should either use two keys, or the single-key approach described by @zaneb where:

```
inspect.metal3.io                                     # re-inspect
inspect.metal3.io: disabled                      # skip inspection
inspect.metal3.io: <data in json format>  # consume data to update status->hardware
```

One question is, do we only support consuming the data in the case where `inspect.metal3.io: disabled` was the previous state?  In that case I guess we'd need to store the state in the status field?

It'd be simpler to consume the data unconditionally, but obviously that does mean some risk a user could overwrite status data we previously collected via inspection (which is a good or bad thing I guess depending on your perspective) :)

Reply:
I'm happy to go with the consensus, @fmuyassarov any preferences?

Reply:
@hardys slightly confused about this discussion, because IIUC what you're proposing is 3 separate keys, the same as in the doc already?

Thinking about this a bit deeper though... I don't think we want to change the semantics of the re-inspection thing, which are that it only does something (flips to the Inspecting state) in the Ready state. And we'll need to be able to set the data when the Host is Provisioned, because we're only going to be able to get it from something running on host and that thing will have to be provisioned. So we probably shouldn't share the same key for setting data and triggering re-inspection after all. Accordingly, I'm amending my suggestion to this:

* `inspect.metal3.io` - move to Inspecting state from Ready
* `inspect.metal3.io=disabled` - don't inspect even in Inspecting
* `inspect.metal3.io/hardwaredetails={...}` - set HardwareDetails if empty and delete annotation

Open question: should we allow users to replace existing HardwareDetails? At any time? Only in the inspecting state after triggering re-inspection?

Reply:
> I'm happy to go with the consensus, @fmuyassarov any preferences?

Hi @hardys , maybe a week would be fine if everyone here is okay with the proposal already. I'm trying also to update the implementation PR for re-inspection. Very sorry for letting that PR to hang for a long time.

Reply:
> @hardys slightly confused about this discussion, because IIUC what you're proposing is 3 separate keys, the same as in the doc already?

Right, personally I think three keys would be fine, but both you and @andfasano expressed a preference for minmising the number of keys, so I'm just trying to bikeshed an interface everyone can agree on ;)
 
> Thinking about this a bit deeper though... I don't think we want to change the semantics of the re-inspection thing, which are that it only does something (flips to the Inspecting state) in the Ready state. And we'll need to be able to set the data when the Host is Provisioned, because we're only going to be able to get it from something running on host and that thing will have to be provisioned. So we probably shouldn't share the same key for setting data and triggering re-inspection after all. Accordingly, I'm amending my suggestion to this:
> 
>     * `inspect.metal3.io` - move to Inspecting state from Ready
> 
>     * `inspect.metal3.io=disabled` - don't inspect even in Inspecting
> 
>     * `inspect.metal3.io/hardwaredetails={...}` - set HardwareDetails if empty and delete annotation

This works for me, my main concern with your initial proposal was conflating the disable/re-inspect action with setting the hardware details.

> Open question: should we allow users to replace existing HardwareDetails? At any time? Only in the inspecting state after triggering re-inspection?

For the use-case I'm currently interested in, we need to update HardwareDetails after the live-iso is booted (since that's the thing collecting inspection data), but in that case we'll set `inspect.metal3.io=disabled` to avoid the time hit of inspecting twice.

So I'd say we need to allow updating `inspect.metal3.io/hardwaredetails`:

  * At any time when `inspect.metal3.io=disabled`
  * On BMH creation - in this case inspection will be skipped, and it's a potentially safer/more constrained interface than the current `baremetalhost.metal3.io/status` ?

In other cases, it probably makes sense to not allow update, since it would mean potential loss of metal3-initiated inspection data, and the user told us to handle inspection?

Reply:
> This works for me, my main concern with your initial proposal was conflating the disable/re-inspect action with setting the hardware details.
:+1: 

> So I'd say we need to allow updating `inspect.metal3.io/hardwaredetails`:
> 
>     * At any time when `inspect.metal3.io=disabled`
> 
>     * On BMH creation - in this case inspection will be skipped, and it's a potentially safer/more constrained interface than the current `baremetalhost.metal3.io/status` ?
> 
Logically speaking since it seems to me that when `inspect.metal3.io=disabled` the inspection is delegated to someone else (responsible to provide the required details in the `hardwaredetails` subkey, I would strictly bind the replacement in the same point (state) where the skipping decision will be done

Reply:
>  I would strictly bind the replacement in the same point (state) where the skipping decision will be done

As I mentioned, there is a requirement to update the hardwaredetails after the live-iso has booted, at which point the BMH will be in the Provisioned state, so we can't only consume the hardwaredetails in the Inspecting state.

That also won't work for externally provisioned hosts since they bypass the Inspecting state, so IMHO it would be simpler to either unconditionally allow `inspect.metal3.io/hardwaredetails` or consume it only when either `inspect.metal3.io=disabled` or there is no existing data in the Status?

Reply:
Another option is we only consume the `inspect.metal3.io/hardwaredetails` annotation when `inspect.metal3.io=disabled` and don't consider the externally provisioned host case.

My motivation in mentioning that is to improve the somewhat awkward pattern where we have to create a BMH but pause reconciling until such time as the status details are added:

https://github.com/metal3-io/baremetal-operator/blob/master/docs/api.md#pausing-reconciliation
https://github.com/openshift/installer/blob/master/pkg/asset/machines/baremetal/hosts.go#L97
https://github.com/openshift/installer/blob/master/data/data/bootstrap/baremetal/files/usr/local/bin/master-bmh-update.sh#L48

@zaneb @dhellmann what are your thoughts, should we consider this new `inspect.metal3.io/hardwaredetails` as a potential alternative for the `baremetalhost.metal3.io/status` interface, or do we need to ensure reconciliation is paused until the hardware details are added in the externally provisioned case?

Reply:
I vote for Consume it only when either inspect.metal3.io=disabled or there is no existing **_HardwareDetails_** data in the Status. That way we won't need the awkward pause thing in OpenShift any more, but we won't overwrite existing data that has come from an actual inspection.

Reply:
Yes, we should use this in OpenShift instead of the mechanism we're using now.

I'm less sure we should apply any rules about when the annotation is consumed. It feels like every time we say "we're only going to want this when..." we find out we're wrong. If the only danger is that a user might overwrite "real" data, can we not just tell them "don't do that if you want things to work properly"? That's how APIs work, right?

If we do want to protect the user, then Zane's proposed handling makes sense as a minimal set of safeguards.

Reply:
I can think of times an unconditional API to overwrite existing hardware status would be useful for debugging/workarounds - so I'm in agreement with @dhellmann that perhaps we should just make `inspect.metal3.io/hardwaredetails` work unconditionally and clearly document that it overwrites any existing data.

Unlike `baremetalhost.metal3.io/status` it only risks removal of the hardware details (at the users request) and can't corrupt other state we rely on that's stored in the Status fields?

Reply:
:+1: for @zaneb proposal, seems to be better as improved safety guards, and it does not look like it will limit the range of applications as previously discussed.

Reply:
Ok it sounds like we've reached consensus on the API but there's still a split of opinion re the hardwaredetails update behavior :sweat_smile: 

Unless there are any further opinions on that, I'll go with the "safer" consume it only if `inspect.metal3.io=disabled` or there is no existing HardwareDetails, then if we find a specific reason to relax that we can consider doing so in future.

Reply:
I'm OK with following the safer approach.